src/vm/types.rs

```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
         Ok(TupleTypeSignature { type_map: type_map })
     }
 
-    pub fn check_valid(&self, name: &str, value: &Value) -> bool {
-        if let Some(expected_type) = self.type_map.get(name) {
-            *expected_type == TypeSignature::type_of(value)
+    pub fn admits(&self, other: &TupleTypeSignature) -> bool {
+        if self.type_map.len() != other.type_map.len() {
+            return false
+        }
+
+        for (name, my_type_sig) in self.type_map.iter() {
+            if let Some(other_type_sig) = other.type_map.get(name) {
+                if !my_type_sig.admits_type(other_type_sig) {
+                    return false
+                }
+            } else {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    pub fn size(&self) -> i128 {
+        let mut name_size: i128 = 0;
+        let mut value_size: i128 = 0;
+        for (name, type_signature) in self.type_map.iter() {
+            // we only accept ascii names, so 1 char = 1 byte.
+            name_size = name_size.checked_add(name.len() as i128).unwrap();
+            value_size = value_size.checked_add(type_signature.size() as i128).unwrap();
+        }
+        let name_total_size = name_size.checked_mul(2).unwrap(); // counts the b-tree size...
+        value_size.checked_add(name_total_size).unwrap()
+    }
+
+    // NOTE: this function mutates self _even if it returns an error_.
```

Comment:
Is doing so necessary?  Is there a risk that this object is left in a partially-computed state?  It might be safer just to clone `self`, apply the `expand_to_admit` method on the copy, and then if everything works, apply the changes to `self`.

Reply:
The expand_to_admit function is only ever called to build a new type, so there's no risk at the moment of using an object that was in the partially computed state.

Reply:
Gotcha.  We'll want to make sure it stays that way.